flux-middle/src/rty/projections.rs

```diff
@@ -1,36 +1,75 @@
+use std::iter;
+
+#[allow(unused_imports)]
 use flux_common::bug;
+use itertools::Itertools;
 use rustc_hash::FxHashMap;
+use rustc_hir::def_id::DefId;
+use rustc_infer::{infer::TyCtxtInferExt, traits::Obligation};
+use rustc_middle::{
+    traits::{ImplSourceUserDefinedData, ObligationCause},
+    ty::{ParamTy, ToPredicate, TraitRef, TyCtxt},
+};
+use rustc_trait_selection::traits::SelectionContext;
 
 use super::{
     fold::{TypeFoldable, TypeFolder, TypeSuperFoldable},
-    AliasKind, AliasTy, BaseTy, ClauseKind, GenericPredicates, Ty, TyKind,
+    AliasKind, AliasTy, BaseTy, ClauseKind, GenericArg, GenericPredicates, Ty, TyKind,
+};
+use crate::{
+    global_env::GlobalEnv,
+    queries::QueryErr,
+    rty::{fold::TypeVisitable, EarlyBinder},
+    rustc::{self},
 };
 
-#[derive(Debug)]
-struct ProjectionTable(FxHashMap<AliasTy, Ty>);
+fn callsite_predicates(genv: &GlobalEnv, def_id: DefId) -> Result<GenericPredicates, QueryErr> {
+    match genv.predicates_of(def_id) {
+        Ok(eb) => Ok(eb.0),
+        Err(e) => Err(e),
+    }
+}
+
+struct ProjectionTable<'sess, 'tcx> {
+    genv: &'sess GlobalEnv<'sess, 'tcx>,
+    def_id: DefId,
+    preds: FxHashMap<AliasTy, Ty>,
+}
 
-impl ProjectionTable {
-    pub fn new(predicates: GenericPredicates) -> Self {
-        let mut res = FxHashMap::default();
+impl<'sess, 'tcx> ProjectionTable<'sess, 'tcx> {
+    fn new(genv: &'sess GlobalEnv<'sess, 'tcx>, def_id: DefId) -> Result<Self, QueryErr> {
+        let predicates = callsite_predicates(genv, def_id)?;
```

Comment:
Always use `skip_binder` (or `instantiate_identity`)

```suggestion
        let predicates = genv.predicates_of(def_id)?.skip_binder();
```